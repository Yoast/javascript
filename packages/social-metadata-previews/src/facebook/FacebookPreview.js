/* External dependencies */
import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

/* Internal dependencies */
import FacebookSiteAndAuthorNames from "./FacebookSiteAndAuthorNames";
import FacebookImage from "./FacebookImage";
import FacebookTitle from "./FacebookTitle";
import FacebookDescription from "./FacebookDescription";

/**
 * Determines the height depending on the mode.
 *
 * @param {string} mode The mode. landscape, square, portrait.
 *
 * @returns {string} The height pixels.
 */
const determineWrapperHeight = ( mode ) => {
	switch ( mode ) {
		case "landscape":
			return "352px";

		case "square":
			return "158px";

		case "portrait":
			return "215px";

		default:
			return "57px";
	}
};

/**
 * Determines the width depending on the mode.
 *
 * @param {string} mode The mode. landscape, square, portrait.
 *
 * @returns {string} The width pixels.
 */
const determineTextContainerWidth = ( mode ) => {
	switch ( mode ) {
		case "landscape":
			return "527x";

		case "square":
			return "369px";

		case "portrait":
			return "369px";

		default:
			return "476px";
	}
};

const FacebookPreviewWrapper = styled.div`
	box-sizing: border-box;
	display: flex;
	flex-direction: ${ props => props.mode === "landscape" ? "column" : "row" };
	background-color: #f2f3f5;
	overflow: hidden;
	width: 527px;
	height: ${ props => determineWrapperHeight( props.mode ) };
`;

const FacebookTextWrapper = styled.div`
	box-sizing: border-box;
	background-color: #f2f3f5;
	margin: 0;
	padding: 10px 12px;
	position: relative;
	border-bottom: ${ props => props.mode === "landscape" ? "" : "1px solid #dddfe2" };
	border-top: ${ props => props.mode === "landscape" ? "" : "1px solid #dddfe2" };
	border-right: ${ props => props.mode === "landscape" ? "" : "1px solid #dddfe2" };
	border: ${ props => props.mode === "landscape" ? "1px solid #dddfe2" : "" };
	display: flex;
	flex-direction: column;
	justify-content: ${ props => props.mode === "landscape" ? "flex-start" : "center" };
	width: ${ props => determineTextContainerWidth( props.mode ) };
	${ props => props.mode === "landscape"  ? "height: 78px;" : "" }
	font-size: 12px;
`;

/**
 * Renders FacebookPreview component.
 *
 * @param {object} props The props.
 *
 * @returns {React.Element} The rendered element.
 */
class FacebookPreview extends Component {
	/**
	 * The constructor.
	 *
	 * @param {Object} props The component's props.
	 */
	constructor( props ) {
		super( props );
		this.state = {
			imageMode: null,
		};
		this.onImageLoaded = this.onImageLoaded.bind( this );
	}

	/**
	 * Retrieves the imageMode from the Facebook image container.
	 *
	 * @param {string} mode The Facebook image mode: landscape, portrait or square.
	 *
	 * @returns {void} Void.
	 */
	onImageLoaded( mode ) {
		this.setState( { imageMode: mode } );
	}

	/**
	 * Renders the FacebookPreview.
	 *
	 * @returns {ReactComponent} Either the PlaceholderImage component, the ErrorImage component or
	 * the TwitterImageContainer.
	 */
	render() {
		const { imageMode } = this.state;
		return (
			<FacebookPreviewWrapper mode={ imageMode }>
				<FacebookImage
					src={ this.props.image }
					alt={ this.props.alt }
					onImageLoaded={ this.onImageLoaded }
				/>
				<FacebookTextWrapper mode={ imageMode }>
					<FacebookSiteAndAuthorNames
						siteName={ this.props.siteName }
						authorName={ this.props.authorName }
						mode={ imageMode }
					/>
					<FacebookTitle title={ this.props.title } />
					<FacebookDescription mode={ imageMode }>
						{ this.props.description }
					</FacebookDescription>
				</FacebookTextWrapper>
			</FacebookPreviewWrapper>
		);
	}
}

FacebookPreview.propTypes = {
	siteName: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired,
	authorName: PropTypes.string,
	description: PropTypes.string,
	image: PropTypes.string,
	alt: PropTypes.string,
};

FacebookPreview.defaultProps = {
	authorName: "",
	description: "",
	alt: "",
	image: "",
};

export default FacebookPreview;
