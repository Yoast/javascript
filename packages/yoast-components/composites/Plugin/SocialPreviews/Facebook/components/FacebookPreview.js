/* External dependencies */
import React, { Fragment } from "react";
import PropTypes from "prop-types";

/* Internal dependencies */
import FacebookSiteName from "./FacebookSiteName";
import FacebookImage from "./FacebookImage";
import FacebookTitle from "./FacebookTitle";

/**
 * Renders a FacebookPreview component.
 *
 * @param {object} props The props.
 *
 * @returns {React.Element} The rendered element.
 */
const FacebookPreview = ( props ) => {
	return (
		<Fragment>
			<FacebookImage src={ props.src } alt={ props.alt } />
			<FacebookSiteName siteName={ props.siteName } />
			<FacebookTitle title={ props.title } />
		</Fragment>
	);
};

FacebookPreview.propTypes = {
	siteName: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired,
	src: PropTypes.string.isRequired,
	alt: PropTypes.string,
};

FacebookPreview.defaultProps = {
	alt: "",
};

export default FacebookPreview;
