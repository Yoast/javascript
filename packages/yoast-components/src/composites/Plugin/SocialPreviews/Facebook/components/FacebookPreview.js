/* External dependencies */
import React, { Fragment } from "react";
import PropTypes from "prop-types";

/* Internal dependencies */
import FacebookSiteName from "./FacebookSiteName";
import FacebookImage from "./FacebookImage";

/**
 * Renders a FacebookPreview component.
 *
 * @param {object} props The props.
 *
 * @returns {React.Element} The rendered element.
 */
const FacebookPreview = ( props ) => {
	return (
		<Fragment>
			<FacebookImage src={ props.src } alt={ props.alt } />
			<FacebookSiteName siteName={ props.siteName } />
		</Fragment>
	);
};

FacebookPreview.propTypes = {
	siteName: PropTypes.string.isRequired,
	src: PropTypes.string.isRequired,
	alt: PropTypes.string,
};

FacebookPreview.defaultProps = {
	alt: "",
};

export default FacebookPreview;
