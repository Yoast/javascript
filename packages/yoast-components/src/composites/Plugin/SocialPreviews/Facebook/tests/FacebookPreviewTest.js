/* External dependencies */
import React from "react";
import renderer from "react-test-renderer";

/* Internal dependencies */
import FacebookPreview from "../components/FacebookPreview";

describe( "FacebookPreview", () => {
	it( "matches the snapshot for a landscape image", () => {
		const component = renderer.create(
			<FacebookPreview siteName="yosat.com" src="https://yoast.com/app/uploads/2015/06/How_to_choose_keywords_FI.png" />
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );
	it( "matches the snapshot for a portrait image", () => {
		const component = renderer.create(
			<FacebookPreview siteName="yoast.com" src="https://yoast.com/app/uploads/2015/09/Author_Joost_x2.png" />
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );
	it( "matches the snapshot for a square image", () => {
		const component = renderer.create(
			<FacebookPreview siteName="yoast.com" src="https://yoast.com/app/uploads/2018/09/avatar_user_1_1537774226.png" />
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );
	it( "matches the snapshot for a too small image", () => {
		const component = renderer.create(
			<FacebookPreview siteName="yoast.com" src="https://yoast.com/app/uploads/2018/11/Logo_TYPO3-250x105.png" />
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );
	it( "matches the snapshot for a faulty image", () => {
		const component = renderer.create(
			<FacebookPreview siteName="yoast.com" src="thisisnoimage" />
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );
} );
