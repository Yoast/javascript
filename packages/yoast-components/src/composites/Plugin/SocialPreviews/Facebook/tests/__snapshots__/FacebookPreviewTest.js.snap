// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacebookPreview matches the snapshot for a faulty image 1`] = `
Array [
  .c0 {
  height: 261px;
  width: 500px;
  background-color: #ddd;
}

<div
    className="c0"
  />,
  .c0 {
  color: #606770;
  font-size: 12px;
  line-height: 11px;
  text-transform: uppercase;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<p
    className="c0"
  >
    yoast.com
  </p>,
]
`;

exports[`FacebookPreview matches the snapshot for a landscape image 1`] = `
Array [
  .c0 {
  height: 261px;
  width: 500px;
  background-color: #ddd;
}

<div
    className="c0"
  />,
  .c0 {
  color: #606770;
  font-size: 12px;
  line-height: 11px;
  text-transform: uppercase;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<p
    className="c0"
  >
    yosat.com
  </p>,
]
`;

exports[`FacebookPreview matches the snapshot for a portrait image 1`] = `
Array [
  .c0 {
  height: 261px;
  width: 500px;
  background-color: #ddd;
}

<div
    className="c0"
  />,
  .c0 {
  color: #606770;
  font-size: 12px;
  line-height: 11px;
  text-transform: uppercase;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<p
    className="c0"
  >
    yoast.com
  </p>,
]
`;

exports[`FacebookPreview matches the snapshot for a square image 1`] = `
Array [
  .c0 {
  height: 261px;
  width: 500px;
  background-color: #ddd;
}

<div
    className="c0"
  />,
  .c0 {
  color: #606770;
  font-size: 12px;
  line-height: 11px;
  text-transform: uppercase;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<p
    className="c0"
  >
    yoast.com
  </p>,
]
`;

exports[`FacebookPreview matches the snapshot for a too small image 1`] = `
Array [
  .c0 {
  height: 261px;
  width: 500px;
  background-color: #ddd;
}

<div
    className="c0"
  />,
  .c0 {
  color: #606770;
  font-size: 12px;
  line-height: 11px;
  text-transform: uppercase;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<p
    className="c0"
  >
    yoast.com
  </p>,
]
`;
